File: web/src/lib/verdict.ts
export type VerdictLabel =
  | "FAKE_DISCOUNT"
  | "BEST_PRICE"
  | "GOOD_DEAL"
  | "FAIR_PRICE"
  | "INSUFFICIENT_DATA";

export interface VerdictResult {
  label: VerdictLabel;
  deal_score: number;
  display: string;
  explanation: string;
  current_price: number;
  avg_30d: number | null;
  all_time_low: number | null;
  all_time_low_date: string | null;
  data_points: number;
  tracking_days: number;
  confidence: number;
}

// A price tracked this long with a known current price is verdict-worthy even
// if it never changed — compact storage keeps stable prices at 1 stored point,
// so we gate on observation *duration*, not change count. A flat price held for
// weeks is itself a strong signal ("no inflated discount").
const MIN_TRACK_DAYS = 14;

function fmt(paisa: number): string {
  const bdt = paisa / 100;
  return `৳${Number.isInteger(bdt) ? bdt.toLocaleString("en-BD") : bdt.toFixed(2)}`;
}

const DISPLAY_EN: Record<VerdictLabel, string> = {
  FAKE_DISCOUNT: "FAKE DISCOUNT",
  BEST_PRICE: "BEST PRICE — ALL-TIME LOW",
  GOOD_DEAL: "GOOD DEAL",
  FAIR_PRICE: "FAIR PRICE",
  INSUFFICIENT_DATA: "TRACKING — NOT ENOUGH DATA YET",
};

const DISPLAY_BN: Record<VerdictLabel, string> = {
  FAKE_DISCOUNT: "ভুয়া ছাড়",
  BEST_PRICE: "সর্বনিম্ন দাম",
  GOOD_DEAL: "ভাল ডিল",
  FAIR_PRICE: "স্বাভাবিক দাম",
  INSUFFICIENT_DATA: "তথ্য সংগ্রহ হচ্ছে",
};

export function getVerdict(
  currentPrice: number,
  prices30d: number[],
  allPrices: number[],
  allTimeLowDate: string | null = null,
  lang: "en" | "bn" = "en",
  trackingDays = 0
): VerdictResult {
  const displayMap = lang === "bn" ? DISPLAY_BN : DISPLAY_EN;
  const dataPoints = allPrices.length;

  // Enough to judge if either: many recent observations, OR we've watched a
  // known price long enough for a flat line to be meaningful.
  const enoughData =
    prices30d.length >= 5 || (trackingDays >= MIN_TRACK_DAYS && currentPrice > 0);

  if (!enoughData) {
    return {
      label: "INSUFFICIENT_DATA",
      deal_score: 5,
      display: displayMap.INSUFFICIENT_DATA,
      explanation:
        lang === "bn"
          ? `আমরা এই পণ্যটি সম্প্রতি ট্র্যাক করা শুরু করেছি।`
          : `We've only just started tracking this product. Check back after a couple of weeks for a price verdict.`,
      current_price: currentPrice,
      avg_30d: null,
      all_time_low: null,
      all_time_low_date: null,
      data_points: dataPoints,
      tracking_days: trackingDays,
      confidence: 0,
    };
  }

  // Stable-price products store one point; fall back to the known current price
  // so averages/lows stay sensible instead of collapsing to NaN.
  const sample30 = prices30d.length ? prices30d : [currentPrice];
  const observed = (allPrices.length ? allPrices : [currentPrice]).concat(currentPrice);
  const avg30d = Math.round(sample30.reduce((a, b) => a + b, 0) / sample30.length);
  const allTimeLow = Math.min(...observed);
  const allTimeHigh = Math.max(...observed);
  const discountFromAvg = avg30d > 0 ? (avg30d - currentPrice) / avg30d : 0;
  // Flat within 1% across all observations = genuinely steady, not a deal/trap.
  const isStable = allTimeHigh - allTimeLow <= Math.max(1, currentPrice * 0.01);

  let label: VerdictLabel;
  let explanation: string;
  let dealScore: number;

  if (isStable) {
    label = "FAIR_PRICE";
    explanation =
      lang === "bn"
        ? `দামটি ${trackingDays} দিনেরও বেশি সময় ধরে ${fmt(currentPrice)} এ স্থির রয়েছে — কোনো বাড়ানো ছাড় নেই।`
        : `Price has held steady at ${fmt(currentPrice)} across ${trackingDays}+ days of tracking — the listed price is genuine, with no inflated "discount".`;
    dealScore = 5;
  } else if (currentPrice > avg30d * 1.05) {
    label = "FAKE_DISCOUNT";
    const overpay = currentPrice - avg30d;
    explanation =
      lang === "bn"
        ? `দামটি ৩০ দিনের গড়ের (${fmt(avg30d)}) চেয়ে ${fmt(overpay)} বেশি।`
        : `Price is ${fmt(overpay)} ABOVE the 30-day average (${fmt(avg30d)}). NOT a good time to buy.`;
    dealScore = Math.max(1, Math.round(3 - (currentPrice / avg30d - 1) * 10));
  } else if (currentPrice <= allTimeLow * 1.02) {
    label = "BEST_PRICE";
    explanation =
      lang === "bn"
        ? `এটি আমাদের ট্র্যাক করা সর্বনিম্ন দাম! সর্বকালীন সর্বনিম্ন: ${fmt(allTimeLow)}।`
        : `Lowest price we've ever tracked! All-time low: ${fmt(allTimeLow)}.`;
    dealScore = 10;
  } else if (discountFromAvg >= 0.1) {
    label = "GOOD_DEAL";
    const pct = Math.round(discountFromAvg * 100);
    explanation =
      lang === "bn"
        ? `দামটি ৩০ দিনের গড় (${fmt(avg30d)}) এর চেয়ে ${pct}% কম।`
        : `Price is ${pct}% below the 30-day average (${fmt(avg30d)}). A good opportunity.`;
    dealScore = Math.min(9, 7 + Math.round(discountFromAvg * 20));
  } else if (discountFromAvg >= 0) {
    label = "FAIR_PRICE";
    explanation =
      lang === "bn"
        ? `দামটি স্বাভাবিক। ৩০ দিনের গড়: ${fmt(avg30d)}।`
        : `Normal price. No special deal right now. 30-day average: ${fmt(avg30d)}.`;
    dealScore = 5;
  } else {
    label = "FAKE_DISCOUNT";
    explanation =
      lang === "bn"
        ? `দামটি গড়ের চেয়ে ${fmt(currentPrice - avg30d)} বেশি। বিজ্ঞাপিত ছাড়টি বিভ্রান্তিকর।`
        : `Price is ${fmt(currentPrice - avg30d)} above the 30-day average (${fmt(avg30d)}). The advertised discount is misleading.`;
    dealScore = Math.max(1, 3 + Math.round(discountFromAvg * 10));
  }

  // Confidence grows with observations or with how long we've watched a price.
  const confidence = Math.min(1, Math.max(dataPoints, trackingDays) / 30);

  return {
    label,
    deal_score: dealScore,
    display: displayMap[label],
    explanation,
    current_price: currentPrice,
    avg_30d: avg30d,
    all_time_low: allTimeLow,
    all_time_low_date: allTimeLowDate,
    data_points: dataPoints,
    tracking_days: trackingDays,
    confidence,
  };
}

File: backend/app/services/verdict.py
"""
DamKoi — Fake Discount Detector (Verdict Engine)

The flagship feature. Determines whether a product's current discount is real
or inflated by comparing against historical price data.

Logic from PRD Section 6.2, with additional edge case handling.
"""

from dataclasses import dataclass
from enum import Enum
from statistics import mean
from typing import List, Optional


class VerdictLabel(str, Enum):
    """Possible verdict outcomes."""
    FAKE_DISCOUNT = "FAKE_DISCOUNT"
    BEST_PRICE = "BEST_PRICE"
    GOOD_DEAL = "GOOD_DEAL"
    FAIR_PRICE = "FAIR_PRICE"
    INSUFFICIENT_DATA = "INSUFFICIENT_DATA"


@dataclass
class Verdict:
    """The complete verdict for a product's current price."""
    label: VerdictLabel
    deal_score: int  # 1-10
    display: str  # e.g., "FAKE DISCOUNT"
    explanation: str  # human-readable reason
    current_price: int  # in paisa
    avg_30d: Optional[int]  # in paisa
    all_time_low: Optional[int]  # in paisa
    all_time_low_date: Optional[str]  # ISO date
    data_points: int
    confidence: float  # 0.0–1.0


# -- Display templates ------------------------------------------------

VERDICT_DISPLAY = {
    VerdictLabel.FAKE_DISCOUNT:     "FAKE DISCOUNT",
    VerdictLabel.BEST_PRICE:        "BEST PRICE -- ALL-TIME LOW",
    VerdictLabel.GOOD_DEAL:         "GOOD DEAL",
    VerdictLabel.FAIR_PRICE:        "FAIR PRICE",
    VerdictLabel.INSUFFICIENT_DATA: "TRACKING -- NOT ENOUGH DATA YET",
}

VERDICT_DISPLAY_BN = {
    VerdictLabel.FAKE_DISCOUNT:     "ভুয়া ছাড়",
    VerdictLabel.BEST_PRICE:        "সর্বনিম্ন দাম",
    VerdictLabel.GOOD_DEAL:         "ভাল ডিল",
    VerdictLabel.FAIR_PRICE:        "স্বাভাবিক দাম",
    VerdictLabel.INSUFFICIENT_DATA: "তথ্য সংগ্রহ হচ্ছে",
}


def _format_price_bdt(paisa: int) -> str:
    """Format paisa amount as BDT string with commas."""
    bdt = paisa / 100
    if bdt == int(bdt):
        return f"৳{int(bdt):,}"
    return f"৳{bdt:,.2f}"


def get_verdict(
    current_price: int,
    prices_last_30_days: List[int],
    all_prices_ever: List[int],
    all_time_low_date: Optional[str] = None,
    lang: str = "en",
) -> Verdict:
    """
    Determine whether the current price represents a real deal.

    Args:
        current_price: Current price in paisa
        prices_last_30_days: List of prices (paisa) observed in last 30 days
        all_prices_ever: List of all prices (paisa) ever recorded
        all_time_low_date: ISO date string of when all-time low was seen
        lang: Language code ("en" or "bn")

    Returns:
        Verdict object with label, score, and explanation
    """
    display_map = VERDICT_DISPLAY_BN if lang == "bn" else VERDICT_DISPLAY
    data_points = len(all_prices_ever)

    # ── Insufficient data check ───────────────────────────────
    if len(prices_last_30_days) < 5:
        if lang == "bn":
            insufficient_explanation = (
                f"আমরা এই পণ্যটি মাত্র {data_points} বার ট্র্যাক করেছি। "
                "নির্ভরযোগ্য বিশ্লেষণের জন্য ১৪ দিনে কমপক্ষে ৫টি মূল্য পয়েন্ট প্রয়োজন।"
            )
        else:
            insufficient_explanation = (
                f"We've only tracked this product {data_points} time(s). "
                "We need at least 5 price points over 14 days to give a reliable verdict."
            )
        return Verdict(
            label=VerdictLabel.INSUFFICIENT_DATA,
            deal_score=5,  # neutral score when we can't judge
            display=display_map[VerdictLabel.INSUFFICIENT_DATA],
            explanation=insufficient_explanation,
            current_price=current_price,
            avg_30d=None,
            all_time_low=None,
            all_time_low_date=None,
            data_points=data_points,
            confidence=0.0,
        )

    avg_30d = int(mean(prices_last_30_days))
    all_time_low = min(all_prices_ever)

    # ── Core verdict logic (from PRD) ─────────────────────────
    discount_from_avg = (avg_30d - current_price) / avg_30d if avg_30d > 0 else 0.0

    if current_price > avg_30d * 1.05:
        label = VerdictLabel.FAKE_DISCOUNT
        overpay = current_price - avg_30d
        if lang == "bn":
            explanation = (
                f"দামটি ৩০ দিনের গড়ের ({_format_price_bdt(avg_30d)}) চেয়ে "
                f"{_format_price_bdt(overpay)} বেশি। এখন কেনার সঠিক সময় নয়।"
            )
        else:
            explanation = (
                f"Price is {_format_price_bdt(overpay)} ABOVE the 30-day average "
                f"({_format_price_bdt(avg_30d)}). This is NOT a good time to buy."
            )
    elif current_price <= all_time_low * 1.02:
        label = VerdictLabel.BEST_PRICE
        if lang == "bn":
            explanation = (
                f"এটি আমাদের ট্র্যাক করা সর্বনিম্ন দাম! "
                f"সর্বকালীন সর্বনিম্ন: {_format_price_bdt(all_time_low)}।"
            )
        else:
            explanation = (
                f"This is the lowest price we've ever tracked! "
                f"All-time low: {_format_price_bdt(all_time_low)}."
            )
    elif discount_from_avg >= 0.10:
        label = VerdictLabel.GOOD_DEAL
        pct = int(discount_from_avg * 100)
        if lang == "bn":
            explanation = (
                f"দামটি ৩০ দিনের গড় ({_format_price_bdt(avg_30d)}) এর চেয়ে {pct}% কম। "
                f"এটি একটি ভালো সুযোগ।"
            )
        else:
            explanation = (
                f"Price is {pct}% below the 30-day average ({_format_price_bdt(avg_30d)}). "
                f"One of the better prices we've seen."
            )
    elif discount_from_avg >= 0.00:
        label = VerdictLabel.FAIR_PRICE
        if lang == "bn":
            explanation = (
                f"দামটি স্বাভাবিক, এখন কোনো বিশেষ ছাড় নেই। "
                f"৩০ দিনের গড়: {_format_price_bdt(avg_30d)}।"
            )
        else:
            explanation = (
                f"Price is normal. No special deal right now. "
                f"30-day average: {_format_price_bdt(avg_30d)}."
            )
    else:
        label = VerdictLabel.FAKE_DISCOUNT
        overpay = current_price - avg_30d
        if lang == "bn":
            explanation = (
                f"দামটি ৩০ দিনের গড় ({_format_price_bdt(avg_30d)}) এর চেয়ে "
                f"{_format_price_bdt(abs(overpay))} বেশি। দেখানো ছাড়টি বিভ্রান্তিকর।"
            )
        else:
            explanation = (
                f"Price is {_format_price_bdt(abs(overpay))} above the 30-day average "
                f"({_format_price_bdt(avg_30d)}). The displayed discount is misleading."
            )

    # ── Calculate deal score (1–10) ───────────────────────────
    deal_score = _calculate_deal_score(current_price, avg_30d, all_time_low, discount_from_avg)

    # ── Confidence based on data density ──────────────────────
    confidence = min(data_points / 30, 1.0)  # max confidence at 30+ data points

    return Verdict(
        label=label,
        deal_score=deal_score,
        display=display_map[label],
        explanation=explanation,
        current_price=current_price,
        avg_30d=avg_30d,
        all_time_low=all_time_low,
        all_time_low_date=all_time_low_date,
        data_points=data_points,
        confidence=confidence,
    )


def _calculate_deal_score(
    current_price: int,
    avg_30d: int,
    all_time_low: int,
    discount_from_avg: float,
) -> int:
    """
    Calculate deal score from 1 to 10.

    Score mapping from PRD:
        9–10: All-time low or very close
        7–8:  Genuinely below average
        5–6:  Normal / fair price
        3–4:  Slightly elevated
        1–2:  Clearly fake / inflated discount
    """
    # Distance from all-time low (0 = at ATL, 1 = far from ATL)
    if all_time_low > 0:
        distance_from_atl = (current_price - all_time_low) / all_time_low
    else:
        distance_from_atl = 0.0

    # At or near all-time low
    if distance_from_atl <= 0.02:
        return 10
    elif distance_from_atl <= 0.05:
        return 9

    # Genuinely below average
    if discount_from_avg >= 0.20:
        return 8
    elif discount_from_avg >= 0.10:
        return 7

    # Normal / fair price
    if discount_from_avg >= 0.02:
        return 6
    elif discount_from_avg >= -0.02:
        return 5

    # Slightly elevated
    if discount_from_avg >= -0.10:
        return 4
    elif discount_from_avg >= -0.15:
        return 3

    # Clearly elevated / fake
    if discount_from_avg >= -0.25:
        return 2
    else:
        return 1
